File: kernel/op-speed/gtsspeed/ops/csrc/wind_swiglu_limit.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */
#include "inc/aclnn_common.h"
#include "pybind.h"

inline std::vector<int64_t> swiglu_grad_infershape(const at::Tensor &x, const c10::optional<at::Tensor> &y, int64_t dim)
{
    if (dim < 0) {
        dim += x.sizes().size();
    }
    TORCH_CHECK(dim < x.sizes().size(), "dim out of range", dim);

    auto input_sizes = x.sizes();
    std::vector<int64_t> output_sizes;
    output_sizes.reserve(input_sizes.size());
    for (size_t i = 0; i < input_sizes.size(); ++i) {
        output_sizes.push_back(input_sizes[i]);
    }
    if (!y.has_value()) {
        output_sizes[dim] /= 2U;
    }
    return output_sizes;
}

// [Add by GTS-AILab] [feature GTSOps] swiglu_limit 前向：x[..2H] -> y[..H]，gate(post-silu)clamp max=limit，up clamp[-limit,limit]
// 调 aclnnSwiGluLimitV2(self, other, dim, limit, highPrecision, out)（v2 op def attr 序：dim,limit,highPrecision）。
// limit<=0 等价普通 swiglu。仅支持合并输入（y 为空）。backend 参数保留以兼容签名，当前仅 v2。
at::Tensor npu_swiglu_limit(const at::Tensor &x, const c10::optional<at::Tensor> &y, int64_t dim, double limit,
                            std::string backend)
{
    TORCH_CHECK(!y.has_value(), "npu_swiglu_limit only supports merged input (y must be None)");
    auto output_sizes = swiglu_grad_infershape(x, y, dim);
    at::Tensor result = at::empty(output_sizes, x.options());
    double limitVal = limit;   // aclnn Float attr ABI = double，必须传具名 double 左值
    bool highPrecision = true;  // 与 v2 op def 默认一致
    bool hasProbs = false;  // 与 v2 op def 默认一致
    at::Tensor out_probs;
    at::Tensor probs;
    ACLNN_CMD(aclnnWindSwiGluLimit, x, y, probs, dim, limitVal, highPrecision, hasProbs, result, out_probs);
    return result;
}

std::tuple<at::Tensor, at::Tensor> npu_swiglu_limit_with_probs(const at::Tensor &x, const c10::optional<at::Tensor> &y, const c10::optional<at::Tensor> &probs, int64_t dim,
                                                               double limit, bool highPrecision, bool hasProbs, std::string backend)
{
    TORCH_CHECK(!y.has_value(), "npu_swiglu_limit_v3 only supports merged input (y must be None)");
    auto output_sizes = swiglu_grad_infershape(x, y, dim);
    at::Tensor result1 = at::empty(output_sizes, x.options());
    at::Tensor result2 = at::empty(output_sizes, x.options());
    double limitVal = limit;   // aclnn Float attr ABI = double，必须传具名 double 左值
    ACLNN_CMD(aclnnWindSwiGluLimit, x, y, probs, dim, limitVal, highPrecision, hasProbs, result1, result2);
    return std::make_tuple(result1, result2);
}

// [Add by GTS-AILab] [feature GTSOps] swiglu_limit 反向：grad_out[..H], x[..2H] -> x_grad[..2H]（与 x 同形）。limit<=0 等价普通 swiglu 反向。
// SwiGluLimitBackwardV2 op def=输入(gradout,self,other OPTIONAL,probs OPTIONAL)+attr(dim,limit)+输出(out1,out2 OPTIONAL)，
//   合并输入下 other 空、out2 空：aclnnSwiGluLimitBackwardV2(gradout, self, other, probs, dim, limit, out1, out2)。
// probs(MoE per-token 权重 [T,1] fp32) 可选：传入则算子内做 grad_h=gradout*probs，
//   省掉外部独立 Mul([*,1])+Cast；probs=None 时与原行为逐字节等价。grad_probs(=sum(grad_u*x)) 仍由调用方外部 reduce。
// backend 参数保留以兼容签名，当前仅 v2。
at::Tensor npu_swiglu_limit_backward(const at::Tensor &grad_output, const at::Tensor &x, int64_t dim, double limit,
                                     const c10::optional<at::Tensor> &probs, std::string backend)
{
    at::Tensor result = at::empty(x.sizes(), x.options());   // x_grad / out1 与 x 同形（全 2H）
    double limitVal = limit;
    c10::optional<at::Tensor> other = c10::nullopt;      // 合并输入：other 为空
    at::Tensor result2 = at::empty({0}, x.options());    // out2 空（无 grad_probs 时）
    ACLNN_CMD(aclnnWindSwiGluLimitBackward, grad_output, x, other, probs, dim, limitVal, result, result2);
    return result;
}

// 重型 probs 版：算子一次出 (grad_x, grad_probs)。grad_probs fp32 = sum_H(grad_out*h)，
// 由算子内逐行 reduce 算（消除外部重算 h + Mul + ReduceSum）。probs 须非空。
std::tuple<at::Tensor, at::Tensor> npu_swiglu_limit_backward_probs(
    const at::Tensor &grad_output, const at::Tensor &x, int64_t dim, double limit,
    const at::Tensor &probs, std::string backend)
{
    at::Tensor result = at::empty(x.sizes(), x.options());   // grad_x 与 x 同形 [T,2H]
    // grad_probs 形状跟随 probs 输入（[T] 或 [T,1] 皆可），autograd 才不会报 grad 与 probs 形状不一致；
    //   dtype 强制 fp32（算子内 fp32 reduce）。必须 at::zeros：切列时原子累加(SetAtomicAdd)要求初值 0。
    //   kernel 按连续 numel 个 float 写 grad_probs，[T]/[T,1] 内存布局相同，故按 probs.sizes() 分配 kernel 照写不误。
    at::Tensor result2 = at::zeros(probs.sizes(), x.options().dtype(at::kFloat));
    double limitVal = limit;
    c10::optional<at::Tensor> other = c10::nullopt;
    c10::optional<at::Tensor> probsOpt = probs;
    ACLNN_CMD(aclnnWindSwiGluLimitBackward, grad_output, x, other, probsOpt, dim, limitVal, result, result2);
    return std::make_tuple(result, result2);
}

File: kernel/op-speed/gtsspeed/ops/wind_swiglu_limit.py
# ----------------------------------------------------------------------------
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ----------------------------------------------------------------------------

import torch
import gtsops

# [Add by GTS-AILab] [feature GTSOps] swiglu_limit 前向+反向（融合 clamp+swiglu，含 clamp 直通掩码）
class NpuSwiGluLimit(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, limit, y=None, dim=-1, backend="v2"):
        # 仅支持合并输入（y=None）；与训练用法一致。backend 默认 v2 融合算子。
        outputs = gtsops.npu_swiglu_limit(x, y, dim, float(limit), backend)
        ctx.save_for_backward(x)
        ctx.dim = dim
        ctx.limit = float(limit)
        ctx.backend = backend   # 存 backend，反向走同版本，前后向成对一致
        return outputs

    @staticmethod
    def backward(ctx, grad_outputs):
        (x,) = ctx.saved_tensors
        # x_grad 与 x 同形（grad_a|grad_b 拼接）。前向 backend → 反向同 backend。
        # probs 用 kwarg 传，避免与新增的 probs 入参位置冲突；此处无 probs(=None)。
        grad_x = gtsops.npu_swiglu_limit_backward(grad_outputs.contiguous(), x, ctx.dim, ctx.limit, backend=ctx.backend)
        # 对应 forward(x, limit, y, dim, backend) 的 5 个入参：只有 x 需要梯度
        return grad_x, None, None, None, None


wind_swiglu_limit = NpuSwiGluLimit.apply


# MoE 专家路径：swiglu_limit + probs 加权（仅反向融合 probs；正向 probs 乘法在外部做）。
# 正向：out = swiglu_limit(x) * probs  —— 正向算子无 probs 输入，probs 乘法是外部 Mul([T,1])，正向不融。
# 反向：grad_x = swiglu_back(grad_out * probs)  —— grad_out*probs 进反向算子（消除外部 Mul([*,1])+Cast）。
#       grad_probs = sum(grad_out * h, -1)  —— 轻型方案：由调用方/上层 routing 反向负责，此函数不返回。
# 注意：forward 内的 *probs 在 autograd.Function 黑盒内，反向全由本 backward 负责，不会 double-count。
class NpuSwiGluLimitProbs(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, y, probs, limit, dim=-1, highprecision=True, hasprobs=True, backend="v3"):
        # x[bf16, Mx2H], probs[float, Mx1] -> output[bf16, MxH]
        outputs = gtsops.npu_swiglu_limit_with_probs(x, y, probs, dim, float(limit), highprecision, hasprobs, backend)
        ctx.save_for_backward(x, probs)
        ctx.dim = dim
        ctx.limit = float(limit)
        ctx.backend = backend
        return outputs

    @staticmethod
    def backward(ctx, grad_without_probs, grad_with_probs):
        x, probs = ctx.saved_tensors
        if grad_with_probs is None:
            return None, None, None, None, None, None, None, None

        # 算子要求 gradout 与 self(x) 同 dtype；上游 grad 可能是 fp32，进算子前 cast 对齐 x.dtype。
        grad_outputs = grad_with_probs.contiguous().to(x.dtype)
        # 重型方案：算子一次出 (grad_x, grad_probs)。grad_x=swiglu_back(grad_out*probs)，
        #   grad_probs=sum_H(grad_out*h) 由算子内逐行 reduce 算（消除外部重算 h + Mul + ReduceSum）。
        # routing 反向需要 probs 梯度（现场 permuted_probs_inputs_detach.grad），返回算子出的 grad_probs。
        grad_x, grad_probs = gtsops.npu_swiglu_limit_backward_probs(
            grad_outputs, x, ctx.dim, ctx.limit, probs, backend=ctx.backend)
        grad_probs = grad_probs.to(probs.dtype)
        if not ctx.needs_input_grad[2]:
            grad_probs = None
        return grad_x, None, grad_probs, None, None, None, None, None


wind_swiglu_limit_probs = NpuSwiGluLimitProbs.apply
